### PR TITLE
Add raw OpenCV building script

### PR DESCRIPTION
**Pull request by zerofivefiveseven**
Add raw CMakeLists.txt
Create some .cpp/.h files

### CODE REVIEW

File: FeatureDetect/src/BuildOpenCV.cmake [hunk 1]
```diff
@@ -0,0 +1,134 @@
+set(PROJECT_BUILD_DIR ${CMAKE_SOURCE_DIR}/build)
+
+set(OPENCV_VERSION 4.6.0)
+set(OPENCV_MODULES core,calib3d,imgproc,imgcodecs,features2d,highgui)
```
Comment (TimPushkin) [inline FeatureDetect/src/BuildOpenCV.cmake]: Add something like `# TODO: remove unused modules` here not to forget about it

File: FeatureDetect/src/BuildOpenCV.cmake [hunk 2]
```diff
@@ -0,0 +1,134 @@
...
+if (CMAKE_TOOLCHAIN_FILE)
+    list(APPEND TOOLCHAIN_ARGS -DCMAKE_TOOLCHAIN_FILE=${CMAKE_TOOLCHAIN_FILE})
+endif ()
+if (CMAKE_C_COMPILER)
+    list(APPEND TOOLCHAIN_ARGS -DCMAKE_C_COMPILER=${CMAKE_C_COMPILER})
+endif ()
+if (CMAKE_CXX_COMPILER)
+    list(APPEND TOOLCHAIN_ARGS -DCMAKE_CXX_COMPILER=${CMAKE_CXX_COMPILER})
+endif ()
+
```
Comment (TimPushkin) [inline FeatureDetect/src/BuildOpenCV.cmake]: It seems like you aren't using the latest version of this script. It has a separate list of Android-related flags here, see [this commit](https://github.com/TimPushkin/SgbmDepthAndroid/commit/746b3b6073f7007b97d2d1c0a7f634854a3680bb) and add it.

File: FeatureDetect/src/BuildOpenCV.cmake [hunk 3]
```diff
@@ -0,0 +1,134 @@
...
+# Configure OpenCV build
+set(
+        OPENCV_CMAKE_ARGS  # https://docs.opencv.org/4.5.5/db/d05/tutorial_config_reference.html
```
Comment (TimPushkin) [inline FeatureDetect/src/BuildOpenCV.cmake]: Update the link to `4.6.0`

File: FeatureDetect/src/BuildOpenCV.cmake [hunk 4]
```diff
@@ -0,0 +1,134 @@
...
+        -DBUILD_ANDROID_PROJECTS=OFF
+        -DBUILD_ANDROID_EXAMPLES=OFF
+        -DBUILD_opencv_apps=OFF
+        -DBUILD_JAVA=OFF
+        -DBUILD_FAT_JAVA_LIB=OFF
+        -DBUILD_KOTLIN_EXTENSIONS=OFF
+        -DBUILD_OBJC=OFF
+        -DBUILD_opencv_python2=OFF
+        -DBUILD_opencv_python3=OFF
+        # Disabled dependencies
```
Comment (TimPushkin) [inline FeatureDetect/src/BuildOpenCV.cmake]: I suggest to remove this for now and add a `TODO` to later disable the unwanted dependecies (when you'll know, which ones are unwanted)

File: FeatureDetect/src/BuildOpenCV.cmake [hunk 5]
```diff
@@ -0,0 +1,134 @@
...
+        -DWITH_MSMF=OFF
+        -DWITH_DIRECTX=OFF
+        -DWITH_VA=OFF
+        -DWITH_VA_INTEL=OFF
+        -DWITH_LAPACK=OFF                 # TODO: make sure it's not used
+        -DWITH_PROTOBUF=OFF               # TODO: make sure it's not used
+        -DWITH_QUIRC=OFF
+        -DWITH_ANDROID_MEDIANDK=OFF
+        -DWITH_ANDROID_NATIVE_CAMERA=OFF
+        # Enabled dependencies
```
Comment (TimPushkin) [inline FeatureDetect/src/BuildOpenCV.cmake]: You can also replace these ones with a `TODO` as you don't know what you actually need now

File: FeatureDetect/src/BuildOpenCV.cmake [hunk 6]
```diff
@@ -0,0 +1,134 @@
...
+        -DWITH_OPENCL=ON
+        -DWITH_OPENVX=ON                  # TODO: install OpenVX and set OPENVX_ROOT
+        -DWITH_EIGEN=ON                   # TODO: install Eigen (and probably use EIGEN_INCLUDE_PATH)
+        -DWITH_IPP=ON
+        # Cross-compilation handling
+        -DCMAKE_TRY_COMPILE_TARGET_TYPE=STATIC_LIBRARY
+        -DANDROID_ABI=${ANDROID_ABI}
+        -DANDROID_ARM_NEON=${ANDROID_ARM_NEON}
```
Comment (TimPushkin) [inline FeatureDetect/src/BuildOpenCV.cmake]: These two will be replaced with `ANDROID_ARGS` when you update the script as I described above

File: FeatureDetect/src/CMakeLists.txt [hunk 7]
```diff
@@ -0,0 +1,19 @@
+cmake_minimum_required(VERSION 3.22.3)
+project(featureDetect)
+
+
+
```
Comment (TimPushkin) [inline FeatureDetect/src/CMakeLists.txt]: Why so many blank lines in this file? I think, one is enough here and also below

File: FeatureDetect/src/external/CMakeLists.txt [hunk 8]
```diff
@@ -0,0 +1 @@
+add_subdirectory(opencv)
```
Comment (TimPushkin) [inline FeatureDetect/src/external/CMakeLists.txt]: Let's add a trailing newline for consistency

File: FeatureDetect/src/CMakeLists.txt [hunk 9]
```diff
@@ -0,0 +1,19 @@
...
+add_subdirectory(source)
+
+add_subdirectory(external)
+
+
+
+
+
+#target_link_libraries(test source)
```
Comment (TimPushkin) [inline FeatureDetect/src/CMakeLists.txt]: Remove this if it isn't needed

File: FeatureDetect/src/external/opencv/CMakeLists.txt [hunk 10]
```diff
@@ -0,0 +1,28 @@
+cmake_minimum_required(VERSION 3.22.3)
+message(STATUS "Configuring OpenCV")
+
+set(OPENCV_VERSION 4.6.0)
+set(OPENCV_PUBLIC_LIBRARIES opencv_core opencv_calib3d opencv_imgproc opencv_imgcodecs opencv_features2d opencv_highgui)
+set(OPENCV_PRIVATE_LIBRARIES  )
```
Comment (TimPushkin) [inline FeatureDetect/src/external/opencv/CMakeLists.txt]: You can just comment this out with a corresponding `TODO` for now

File: FeatureDetect/src/external/opencv/CMakeLists.txt [hunk 11]
```diff
@@ -0,0 +1,28 @@
...
+        if (OpenCV_FOUND)
+            break()
+        endif ()
+    endforeach ()
+endif ()
+
+find_package(OpenCV ${OPENCV_VERSION} REQUIRED ${OPENCV_PUBLIC_LIBRARIES} ${OPENCV_PRIVATE_LIBRARIES})
+
+target_link_libraries(${PROJECT_NAME} PUBLIC ${OPENCV_PUBLIC_LIBRARIES})
+#target_link_libraries(${PROJECT_NAME} PRIVATE ${OPENCV_PRIVATE_LIBRARIES})
```
Comment (TimPushkin) [inline FeatureDetect/src/external/opencv/CMakeLists.txt]: Add a space between `#` and the first letter, please. This is the style of other comments and makes them more readable.

File: FeatureDetect/src/samples/test.cpp [hunk 12]
```diff
@@ -0,0 +1,7 @@
+#include "featurelib.h"
+
+int main()
+{
```
Comment (TimPushkin) [inline FeatureDetect/src/samples/test.cpp]: Let's pick a codestyle that is supported by some linter and stick to it.

For example, if we are writing for Android mainly, we can use [this one](https://google.github.io/styleguide/cppguide.html) from Google. It is supported in [cpplint](https://github.com/cpplint/cpplint).

File: FeatureDetect/src/samples/test.cpp [hunk 13]
```diff
@@ -0,0 +1,7 @@
...
+featurelib instance();
+    return 0;
+}
```
Comment (TimPushkin) [inline FeatureDetect/src/samples/test.cpp]: Add a trailing newline

File: FeatureDetect/src/source/CMakeLists.txt [hunk 14]
```diff
@@ -0,0 +1,13 @@
+cmake_minimum_required(VERSION 3.22.3)
+#project(source)
+#set(MY_SOURCE_LIB featurelib.cpp)
+#add_library(source STATIC ${MY_SOURCE_LIB})
+#include_directories(source)
```
Comment (TimPushkin) [inline FeatureDetect/src/source/CMakeLists.txt]: Remove these comments (and the comment in the end of the file) if they are not needed anymore

File: FeatureDetect/src/source/CMakeLists.txt [hunk 15]
```diff
@@ -0,0 +1,13 @@
...
+set(SOURCE_FLS featurelib.cpp)
```
Comment (TimPushkin) [inline FeatureDetect/src/source/CMakeLists.txt]: Don't shorten words like this where unnecessary -- I'm speaking about `FLS`

File: FeatureDetect/src/source/featurelib.cpp [hunk 16]
```diff
@@ -0,0 +1,8 @@
+
+
```
Comment (TimPushkin) [inline FeatureDetect/src/source/featurelib.cpp]: Why so many newlines here and below?

File: FeatureDetect/src/source/featurelib.h [hunk 17]
```diff
@@ -0,0 +1,27 @@
+
```
Comment (TimPushkin) [inline FeatureDetect/src/source/featurelib.h]: I think, this newline is not needed

File: FeatureDetect/src/source/featurelib.h [hunk 18]
```diff
@@ -0,0 +1,27 @@
...
+#include <opencv2/core/ocl.hpp>
+
+class featurelib  {
+public:
+    featurelib();
+private:
+    cv::Ptr<cv::Feature2D> sift = cv::SIFT::create();
+};
+
+
```
Comment (TimPushkin) [inline FeatureDetect/src/source/featurelib.h]: Too many newlines

File: FeatureDetect/src/source/featurelib.h [hunk 19]
```diff
@@ -0,0 +1,27 @@
...
+#include <opencv2/objdetect.hpp>
+#include <opencv2/core/core.hpp>
+#include <opencv2/highgui/highgui.hpp>
+#include <opencv2/features2d.hpp>
+//#include <opencv2/xfeatures2d.hpp>
+//#include <opencv2/xfeatures2d/nonfree.hpp>
+#include <opencv2/calib3d.hpp>
+#include <opencv2/imgproc.hpp>
+#include <opencv2/core/utility.hpp>
+#include <opencv2/core/ocl.hpp>
```
Comment (TimPushkin) [inline FeatureDetect/src/source/featurelib.h]: Why so many imports? It seems like they are unused.

**Review by TimPushkin — CHANGES_REQUESTED** [review]
1. Replace the `FeatureDetect/src/samples/starry_night.png` artwork with something else because it might be copyrighted. Pick something small, neutral, and having more obvious keypoints (you can take a picture of something parallelepiped-like with simple texture, for example). And do it with `git commit --amend` and `git push --force` so that this artwork doesn't stay in repo's history. You can also read about force-pushing and why it is dangerous. **Start with this task before doing anything else**.
2. Let's use directory layout described [here](https://api.csswg.org/bikeshed/?force=1&url=https://raw.githubusercontent.com/vector-of-bool/pitchfork/develop/data/spec.bs), and let the top level directory be the root, not `src` like it is now.
3. CMake required version `3.22.3` is too high. Make it not greater than `3.18.1` because it is the one shipped with Android Studio.

File: FeatureDetect/src/BuildOpenCV.cmake (hunk 2 above)
Comment (zerofivefiveseven) [reply FeatureDetect/src/BuildOpenCV.cmake]: done

File: FeatureDetect/src/BuildOpenCV.cmake (hunk 1 above)
Comment (zerofivefiveseven) [reply FeatureDetect/src/BuildOpenCV.cmake]: done

File: FeatureDetect/src/BuildOpenCV.cmake (hunk 3 above)
Comment (zerofivefiveseven) [reply FeatureDetect/src/BuildOpenCV.cmake]: done

File: FeatureDetect/src/BuildOpenCV.cmake (hunk 2 above)
Comment (zerofivefiveseven) [reply FeatureDetect/src/BuildOpenCV.cmake]: я забыл запуллить просто :)

File: FeatureDetect/src/BuildOpenCV.cmake (hunk 4 above)
Comment (zerofivefiveseven) [reply FeatureDetect/src/BuildOpenCV.cmake]: done, moved code to comments

File: FeatureDetect/src/BuildOpenCV.cmake (hunk 5 above)
Comment (zerofivefiveseven) [reply FeatureDetect/src/BuildOpenCV.cmake]: done

File: FeatureDetect/src/BuildOpenCV.cmake (hunk 6 above)
Comment (zerofivefiveseven) [reply FeatureDetect/src/BuildOpenCV.cmake]: done

File: FeatureDetect/src/CMakeLists.txt (hunk 7 above)
Comment (zerofivefiveseven) [reply FeatureDetect/src/CMakeLists.txt]: done

File: FeatureDetect/src/external/CMakeLists.txt (hunk 8 above)
Comment (zerofivefiveseven) [reply FeatureDetect/src/external/CMakeLists.txt]: done

File: FeatureDetect/src/CMakeLists.txt (hunk 9 above)
Comment (zerofivefiveseven) [reply FeatureDetect/src/CMakeLists.txt]: done

File: FeatureDetect/src/external/opencv/CMakeLists.txt (hunk 10 above)
Comment (zerofivefiveseven) [reply FeatureDetect/src/external/opencv/CMakeLists.txt]: done

File: FeatureDetect/src/external/opencv/CMakeLists.txt (hunk 11 above)
Comment (zerofivefiveseven) [reply FeatureDetect/src/external/opencv/CMakeLists.txt]: Thanks, for the codestyle hint. done

File: FeatureDetect/src/samples/test.cpp (hunk 12 above)
Comment (zerofivefiveseven) [reply FeatureDetect/src/samples/test.cpp]: i install that, done

File: FeatureDetect/src/samples/test.cpp (hunk 13 above)
Comment (zerofivefiveseven) [reply FeatureDetect/src/samples/test.cpp]: done

File: FeatureDetect/src/source/CMakeLists.txt (hunk 14 above)
Comment (zerofivefiveseven) [reply FeatureDetect/src/source/CMakeLists.txt]: done

File: FeatureDetect/src/source/featurelib.h (hunk 19 above)
Comment (zerofivefiveseven) [reply FeatureDetect/src/source/featurelib.h]: done

File: FeatureDetect/src/source/CMakeLists.txt (hunk 15 above)
Comment (zerofivefiveseven) [reply FeatureDetect/src/source/CMakeLists.txt]: done

File: FeatureDetect/src/source/featurelib.cpp (hunk 16 above)
Comment (zerofivefiveseven) [reply FeatureDetect/src/source/featurelib.cpp]: done

File: FeatureDetect/src/source/featurelib.h (hunk 17 above)
Comment (zerofivefiveseven) [reply FeatureDetect/src/source/featurelib.h]: done

File: FeatureDetect/src/source/featurelib.h (hunk 18 above)
Comment (zerofivefiveseven) [reply FeatureDetect/src/source/featurelib.h]: done

File: FeatureDetect/src/BuildOpenCV.cmake (hunk 5 above)
Comment (TimPushkin) [reply FeatureDetect/src/BuildOpenCV.cmake]: No, I meant that you can also replace the ones under the `Enabled dependencies` line with a `TODO`

File: FeatureDetect/root/CMakeLists.txt [hunk 20]
```diff
@@ -0,0 +1,12 @@
...
+
+add_subdirectory(examples)
+
+add_subdirectory(source)
+
+add_subdirectory(external)
+
+
+
+
```
Comment (TimPushkin) [inline FeatureDetect/root/CMakeLists.txt]: Still a lot of blank lines here

File: FeatureDetect/src/CMakeLists.txt (hunk 7 above)
Comment (TimPushkin) [reply FeatureDetect/src/CMakeLists.txt]: Nope, this file still has a lot of blank lines in the end

File: FeatureDetect/root/examples/test.cpp [hunk 21]
```diff
@@ -0,0 +1,6 @@
+#include "../source/featurelib.h"
+
+int main() {
+featurelib instance();
```
Comment (TimPushkin) [inline FeatureDetect/root/examples/test.cpp]: Is it really ok for the linter that this line has no indentation?

File: FeatureDetect/root/source/featurelib.h [hunk 22]
```diff
@@ -0,0 +1,18 @@
...
+#include <opencv2/imgproc.hpp>
+
+class featurelib  {
+public:
+     featurelib();
+private:
+     cv::Ptr<cv::Feature2D> sift = cv::SIFT::create();
+};
+
+#endif //FEATUREDETECT_FEATURELIB_H_
```
Comment (TimPushkin) [inline FeatureDetect/root/source/featurelib.h]: Add a space between `//` and `FEATUREDETECT_FEATURELIB_H_` -- as I said before, it makes the comments more readable

**Review by TimPushkin — CHANGES_REQUESTED** [review]
1. Some conversations from the previous review stay unresolved, so fix them first.
2. The church picture you added -- are you sure it isn't also copyrighted? If you aren't sure (for this you need to have a license agreement from it's authors which states that it's free for usage without even stating the authorship) replace it with something again the same way you did with the previous artwork (through force-pushing, though, it'll be harder now). I'll advice you to just add something of what you are the author to spare us from possible legal problems.
3. When I said about root being the top-level directory I didn't meant a directory called `root` literally :) I meant that you should work in the root directory of the repository as we'll change the top-level file structure after merging your part and the Android part. So you can just remove `FeatureDetect/root` prefix from all the files and get what I want.
4. The file structure I wanted you to use doesn't have `source` directory -- it is `src`. And `featurelib.h` should reside in `include` directory.

File: FeatureDetect/root/examples/test.cpp (hunk 21 above)
Comment (zerofivefiveseven) [reply FeatureDetect/root/examples/test.cpp]: paul@paul-VivoBook-ASUSLaptop-X515EA-R565EA:~/CLionProjects/FeatureDetectAndroid/examples$ cpplint test.cpp
test.cpp:0:  No copyright message found.  You should have a line: "Copyright [year] <Copyright Owner>"  [legal/copyright] [5]
Done processing test.cpp
Total errors found: 1
(base) paul@paul-VivoBook-ASUSLaptop-X515EA-R565EA:~/CLionProjects/FeatureDetectAndroid/examples$ cat test.cpp
#include "../source/featurelib.h"

int main() {
featurelib instance();
    return 0;

Comment (zerofivefiveseven) [reply FeatureDetect/root/examples/test.cpp]: yep

File: FeatureDetect/root/source/featurelib.h (hunk 22 above)
Comment (zerofivefiveseven) [reply FeatureDetect/root/source/featurelib.h]: done

File: FeatureDetect/root/examples/test.cpp (hunk 21 above)
Comment (TimPushkin) [reply FeatureDetect/root/examples/test.cpp]: Well, then fix it by hand as it doesn't comply with [the code style](https://google.github.io/styleguide/cppguide.html#Function_Declarations_and_Definitions)

File: examples/test.cpp [hunk 23]
```diff
@@ -0,0 +1,6 @@
+#include "../include/featurelib.h"
+
+int main() {
+featurelib instance();
```
Comment (TimPushkin) [inline examples/test.cpp]: Fix this line's indentation. Also, CLion warns me about these parantheses and I think it's right, so remove them.

File: include/featurelib.h [hunk 24]
```diff
@@ -0,0 +1,18 @@
...
+#define FEATUREDETECT_FEATURELIB_H_
+
+#include <vector>
+#include <iostream>
+#include <opencv2/opencv.hpp>
+#include <opencv2/core/core.hpp>
+#include <opencv2/features2d.hpp>
+#include <opencv2/imgproc.hpp>
+
+class featurelib  {
```
Comment (TimPushkin) [inline include/featurelib.h]: There is an extra space between `featurelib` and `{` -- CLion should fix such issues when you press `Ctrl + Alt + L`

File: include/featurelib.h [hunk 25]
```diff
@@ -0,0 +1,18 @@
...
+public:
```
Comment (TimPushkin) [inline include/featurelib.h]: `public`/`private` constructs should be indented with one space, according to [our codestyle](https://google.github.io/styleguide/cppguide.html#Class_Format). You can setup CLion to also do it for you.

File: src/CMakeLists.txt [hunk 26]
```diff
@@ -0,0 +1,7 @@
+cmake_minimum_required(VERSION 3.18.1)
+
+set(SOURCE_FILES featurelib.cpp)
+
+add_library(${PROJECT_NAME} SHARED ${SOURCE_FILES})
+target_include_directories(${PROJECT_NAME} PUBLIC ${CMAKE_SOURCE_DIR}/include)
+
```
Comment (TimPushkin) [inline src/CMakeLists.txt]: Remove this extra empty line

**Review by TimPushkin — CHANGES_REQUESTED** [review]
Some more code style issues

File: FeatureDetect/root/examples/test.cpp (hunk 21 above)
Comment (zerofivefiveseven) [reply FeatureDetect/root/examples/test.cpp]: done

File: include/featurelib.h (hunk 24 above)
Comment (zerofivefiveseven) [reply include/featurelib.h]: done

File: examples/test.cpp (hunk 23 above)
Comment (zerofivefiveseven) [reply examples/test.cpp]: done

File: src/CMakeLists.txt (hunk 26 above)
Comment (zerofivefiveseven) [reply src/CMakeLists.txt]: done

File: include/featurelib.h (hunk 25 above)
Comment (zerofivefiveseven) [reply include/featurelib.h]: done

File: examples/test.cpp [hunk 27]
```diff
@@ -0,0 +1,6 @@
+#include "../include/featurelib.h"
+
+int main() {
+    featurelib instance;
+    return 0;
```
Comment (TimPushkin) [inline examples/test.cpp]: Seems like these two lines have wrong indentations. Should be 2 spaces, according to the code style. Just configure Google's code style in CLion and press `Ctrl+Alt+L` in every file.

File: external/opencv/CMakeLists.txt [hunk 28]
```diff
@@ -0,0 +1,28 @@
+cmake_minimum_required(VERSION 3.18.1)
```
Comment (TimPushkin) [inline external/opencv/CMakeLists.txt]: This line is not needed here as it is present in the top-level CMake

File: src/CMakeLists.txt [hunk 29]
```diff
@@ -0,0 +1,6 @@
+cmake_minimum_required(VERSION 3.18.1)
```
Comment (TimPushkin) [inline src/CMakeLists.txt]: Not needed here too

**Review by TimPushkin — CHANGES_REQUESTED** [review]
Fix these and the one left from the previos review

File: examples/test.cpp (hunk 27 above)
Comment (zerofivefiveseven) [reply examples/test.cpp]: done

File: external/opencv/CMakeLists.txt (hunk 28 above)
Comment (zerofivefiveseven) [reply external/opencv/CMakeLists.txt]: done

File: src/CMakeLists.txt (hunk 29 above)
Comment (zerofivefiveseven) [reply src/CMakeLists.txt]: done

**Review by TimPushkin — APPROVED** [review]
🎉

**Pull request merged**